internal/declcfg/declcfg_to_model.go

```diff
@@ -0,0 +1,113 @@
+package declcfg
+
+import (
+	"fmt"
+
+	"github.com/operator-framework/operator-registry/internal/model"
+	"github.com/operator-framework/operator-registry/internal/property"
+)
+
+func ConvertToModel(cfg DeclarativeConfig) (model.Model, error) {
+	mpkgs := model.Model{}
+	defaultChannels := map[string]string{}
+	for _, p := range cfg.Packages {
+		mpkg := &model.Package{
+			Name:        p.Name,
+			Description: p.Description,
+			Channels:    map[string]*model.Channel{},
+		}
+		if p.Icon != nil {
+			mpkg.Icon = &model.Icon{
+				Data:      p.Icon.Data,
+				MediaType: p.Icon.MediaType,
+			}
+		}
+		defaultChannels[p.Name] = p.DefaultChannel
+		mpkgs[p.Name] = mpkg
+	}
+
+	for _, b := range cfg.Bundles {
+		defaultChannelName := defaultChannels[b.Package]
+		if b.Package == "" {
+			return nil, fmt.Errorf("package name must be set for bundle %q", b.Name)
+		}
+		mpkg, ok := mpkgs[b.Package]
+		if !ok {
+			return nil, fmt.Errorf("unknown package %q for bundle %q", b.Package, b.Name)
+		}
+
+		props, err := parseProperties(b.Properties)
+		if err != nil {
+			return nil, fmt.Errorf("parse properties for bundle %q: %v", b.Name, err)
+		}
+
+		if b.Package != props.Packages[0].PackageName {
+			return nil, fmt.Errorf("package %q does not match %q property %q", b.Package, property.TypePackage, props.Packages[0].PackageName)
+		}
+
+		for _, bundleChannel := range props.Channels {
+			pkgChannel, ok := mpkg.Channels[bundleChannel.Name]
+			if !ok {
+				pkgChannel = &model.Channel{
+					Package: mpkg,
+					Name:    bundleChannel.Name,
+					Bundles: map[string]*model.Bundle{},
+				}
+				if bundleChannel.Name == defaultChannelName {
+					mpkg.DefaultChannel = pkgChannel
+				}
+				mpkg.Channels[bundleChannel.Name] = pkgChannel
+			}
+			pkgChannel.Bundles[b.Name] = &model.Bundle{
+				Package:       mpkg,
+				Channel:       pkgChannel,
+				Name:          b.Name,
+				Image:         b.Image,
+				Replaces:      bundleChannel.Replaces,
+				Skips:         skipsToStrings(props.Skips),
+				Properties:    b.Properties,
+				RelatedImages: relatedImagesToModelRelatedImages(b.RelatedImages),
+				CsvJSON:       b.CsvJSON,
+				Objects:       b.Objects,
+			}
+		}
+	}
+
+	for _, mpkg := range mpkgs {
```

Comment:
Will need to revisit this part of code as this will add an empty channel entry to the channel list which will error out later because we don't allow an invalid/empty channel. This happens when `defauftChannel` is missing from package blob. The error will complain about an empty channel which is not entirely accurate when it should be missing default channel. Add this comment here for a record and will address this later.